File: HashMap/Insert Delete Get Random O(1)/RandomizedSet.cpp
#include <cstdlib>
#include <unordered_map>
#include <vector>

class RandomizedSet{
public:
    unordered_map<int, int> m;  //to store the valse and their indices in the list
    vector<int> list;

    RandomizedSet() {} //initalizes an empty RandomizedSet

    bool insert(int val){
        //if the value alredy exist in the map, return false
        if(m.find(val) != m.end()){
            return false;
        }

        //insert the value in the map with its index in the list
        m[val] = list.size();
        //add the value in the list from the end
        list.push_back(val);

        return true;
    }

    bool remove(int val){
        //if the value do not exist, return flase
        if(m.find(val) == m.end()){
            return true;
        }

        //get the index of the vale to be removed from the map
        int remidx = m[val];
        //remove the vale from the map
        m.erase(val);

        //if that value is the last element in the list, remove directly
        if(remidx == list.size()-1){
            list.pop_back();
            return true;
        }

        //otherwise, swap that value with the last element in the list
        list[remidx] = list.back();
        //remove the last element
        list.pop_back();

        //update the map with the new index of the swapped value
        m[list[remidx]] = remidx;

        return true;
    }

    int getRandom(){
        //generate a random number from 0 to the size of the list
        int randomidx = rand() % list.size();

        //return the value corresponding to the random number(index) in the list
        return list[randomidx];
    }
};

File: HashMap/Insert Delete Get Random O(1)/RandomizedSet.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Random;

class Randomizedset {
    // use hashmap to store values and their indices in the list
    HashMap<Integer, Integer> m;
    List<Integer> list;
    Random r;

    public Randomizedset(){
        m = new HashMap<>();
        list = new ArrayList<>();
        r = new Random();
    } 

    public boolean insert(int val){
        //if the value already exist in the map, we don't need to insert again -> 
        // return false
        if(m.containsKey(val)){
            return false;
        }

        //insert the value in the map, with it's index in the list
        m.put(val, list.size());
        //add that value at the end of the list
        list.add(val);

        return true;
    }

    public boolean delete(int val){
        //if the value do not exist, return false
        if(!m.containsKey(val)){
            return false;
        }
        
        //get the index of the value to be deleted
        int remidx = m.get(val);
        //delete that value from the hashmap
        m.remove(val);

        //if the value is the last element in the list, simply remove it
        if(remidx == list.size()-1){
            list.remove(list.size()-1);
            return true;
        }

        //otherwise, swap that element with the last element in the list
        list.set(remidx, list.get(list.size()-1));
        //remove the list element
        list.remove(list.size()-1);

        //update the hashmap to reflect the new index of the last element(which is swapped)
        m.put(list.get(remidx), remidx);

        return true;
    }

    public int getRandom(){
        //generate a random integer b/w 0 and list.size()-1
        int randomidx = r.nextInt(list.size());

        //return the random integer as index of the list 
        return list.get(randomidx);
    }
}
